undefined reference to `__device_dts_ord_20' When building with board hifive_unmatched on flash_shell samples

**Describe the bug**
I was trying to build the flash_shell samples using the board hifive_unmatched and the compilation failed with 

```
/p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/drivers/flash/spi_nor.c:1230: 
undefined reference to `__device_dts_ord_20'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

**Board**
hifive_unmatched 

**What have you tried to diagnose or workaround this issue?**
- I verified the DT_DRV_COMPAT has named correctly
- I verified that the dt/bindings is in placed

**To Reproduce**
Steps to reproduce the behavior:
1. west init zephyrproject
2. cd zephyrproject
3. west update
4. cd zephyr
5. west build -b hifive_unmatched samples/drivers/flash_shell --pristine
6. see error

**Expected behavior**
Build success without error

**Impact**
(Showstoppers) This is blocking the PR https://github.com/zephyrproject-rtos/zephyr/pull/46443

**Logs and console output**
```
[09/07/2022 - 07:19.01] [boonkhai.boonkhai-container-55d96b56d8-fgzrj]
/p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr
$ west build -b hifive_unmatched samples/drivers/flash_shell --pristine
-- west build: generating a build system
Loading Zephyr default modules (Zephyr repository).
-- Application: /p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/samples/drivers/flash_shell
-- Found Python3: /build/zephyrproject/.venv/bin/python3.6 (found suitable exact version "3.6.15") found components: Interpreter 
-- Cache files will be written to: /nfs/site/home/boonkhai/.cache/zephyr
-- Zephyr version: 3.1.99 (/p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr)
-- Found west (found suitable version "0.12.0", minimum required is "0.7.1")
-- Board: hifive_unmatched
-- Found host-tools: zephyr 0.13.2 (/build/zephyr-sdk-0.13.2)
-- Found dtc: /build/zephyr-sdk-0.13.2/sysroots/x86_64-pokysdk-linux/usr/bin/dtc (found suitable version "1.6.0", minimum required is "1.4.6")
-- Found toolchain: zephyr 0.13.2 (/build/zephyr-sdk-0.13.2)
-- Found BOARD.dts: /p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/boards/riscv/hifive_unmatched/hifive_unmatched.dts
-- Generated zephyr.dts: /p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/build/zephyr/zephyr.dts
-- Generated devicetree_unfixed.h: /p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/build/zephyr/include/generated/devicetree_unfixed.h
-- Generated device_extern.h: /p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/build/zephyr/include/generated/device_extern.h
-- Including generated dts.cmake file: /p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/build/zephyr/dts.cmake
Parsing /p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/samples/drivers/flash_shell/Kconfig
Loaded configuration '/p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/boards/riscv/hifive_unmatched/hifive_unmatched_defconfig'
Merged configuration '/p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/samples/drivers/flash_shell/prj.conf'
Configuration saved to '/p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/build/zephyr/.config'
Kconfig header saved to '/p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/build/zephyr/include/generated/autoconf.h'
-- The C compiler identification is GNU 10.3.0
-- The CXX compiler identification is GNU 10.3.0
-- The ASM compiler identification is GNU
-- Found assembler: /build/zephyr-sdk-0.13.2/riscv64-zephyr-elf/bin/riscv64-zephyr-elf-gcc
-- Configuring done
-- Generating done
-- Build files have been written to: /p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/build
-- west build: building application
[1/163] Preparing syscall dependency handling

[3/163] Generating include/generated/version.h
-- Zephyr version: 3.1.99 (/p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr), build: zephyr-v3.1.0-1882-gb161554ee862
[49/163] Building C object CMakeFiles/app.dir/src/main.c.obj
/p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/samples/drivers/flash_shell/src/main.c: In function 'do_write_unaligned':
/p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/samples/drivers/flash_shell/src/main.c:304:23: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
  304 |   off_t data_offset = (off_t)buf;
      |                       ^
/p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/samples/drivers/flash_shell/src/main.c:317:7: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
  317 |       (const void *)data_offset,
      |       ^
[153/163] Linking C executable zephyr/zephyr_pre0.elf
FAILED: zephyr/zephyr_pre0.elf zephyr/zephyr_pre0.map 
: && ccache /build/zephyr-sdk-0.13.2/riscv64-zephyr-elf/bin/riscv64-zephyr-elf-gcc   zephyr/CMakeFiles/zephyr_pre0.dir/misc/empty_file.c.obj -o zephyr/zephyr_pre0.elf  -fuse-ld=bfd  -Wl,-T  zephyr/linker_zephyr_pre0.cmd  -Wl,-Map=/p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/build/zephyr/zephyr_pre0.map  -Wl,--whole-archive  app/libapp.a  zephyr/libzephyr.a  zephyr/arch/common/libarch__common.a  zephyr/arch/arch/riscv/core/libarch__riscv__core.a  zephyr/lib/libc/minimal/liblib__libc__minimal.a  zephyr/lib/posix/liblib__posix.a  zephyr/drivers/interrupt_controller/libdrivers__interrupt_controller.a  zephyr/drivers/console/libdrivers__console.a  zephyr/drivers/spi/libdrivers__spi.a  zephyr/drivers/flash/libdrivers__flash.a  zephyr/drivers/serial/libdrivers__serial.a  zephyr/drivers/timer/libdrivers__timer.a  -Wl,--no-whole-archive  zephyr/kernel/libkernel.a  zephyr/CMakeFiles/offsets.dir/./arch/riscv/core/offsets/offsets.c.obj  -L"/build/zephyr-sdk-0.13.2/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/10.3.0/rv64imac/lp64/medany"  -L/p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/build/zephyr  -lgcc  zephyr/arch/common/libisr_tables.a  -no-pie  -mcmodel=medany  -mabi=lp64  -march=rv64imac  -Wl,--gc-sections  -Wl,--build-id=none  -Wl,--sort-common=descending  -Wl,--sort-section=alignment  -Wl,-u,_OffsetAbsSyms  -Wl,-u,_ConfigAbsSyms  -nostdlib  -static  -Wl,-X  -Wl,-N  -Wl,--orphan-handling=warn && cd /p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/build/zephyr && /usr/local/bin/cmake -E echo
/build/zephyr-sdk-0.13.2/riscv64-zephyr-elf/bin/../lib/gcc/riscv64-zephyr-elf/10.3.0/../../../../riscv64-zephyr-elf/bin/ld.bfd: zephyr/drivers/flash/libdrivers__flash.a(spi_nor.c.obj):/p/psg/swip/w/boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/drivers/flash/spi_nor.c:1230: undefined reference to `__device_dts_ord_20'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/local/bin/cmake --build /nfs/png/disks/swuser_work_boonkhai/git/boonkhai_github/zephyr-from-mainline/zephyrproject/zephyr/build

```

**Environment (please complete the following information):**
- Linux 
- Zephyr SDK
- zephyr-sdk-0.13.2

--- a/boards/riscv/hifive1/hifive1.dts
+++ b/boards/riscv/hifive1/hifive1.dts
@@ -50,12 +50,12 @@
 	pinctrl-names = "default";
 };
 
+/* disabled (used by Flash ROM by default) */
 &spi0 {
-	status = "okay";
-
 	reg = <0x10014000 0x1000 0x20400000 0xc00000>;
 	flash0: flash@0 {
 		compatible = "issi,is25lp128", "jedec,spi-nor";
+		status = "disabled";
 		size = <134217728>;
 		label = "FLASH0";
 		jedec-id = [96 60 18];

--- a/boards/riscv/hifive1_revb/hifive1_revb.dts
+++ b/boards/riscv/hifive1_revb/hifive1_revb.dts
@@ -92,12 +92,12 @@
 	current-speed = <115200>;
 };
 
+/* disabled (used by Flash ROM by default) */
 &spi0 {
-	status = "okay";
-
 	reg = <0x10014000 0x1000 0x20010000 0x3c0900>;
 	flash0: flash@0 {
 		compatible = "issi,is25lp128", "jedec,spi-nor";
+		status = "disabled";
 		size = <134217728>;
 		label = "FLASH0";
 		jedec-id = [96 60 18];

--- a/boards/riscv/hifive_unleashed/hifive_unleashed.dts
+++ b/boards/riscv/hifive_unleashed/hifive_unleashed.dts
@@ -46,12 +46,12 @@
 	current-speed = <115200>;
 };
 
+/* disabled (used by Flash ROM by default) */
 &spi0 {
-	status = "okay";
-
 	reg = <0x10040000 0x1000 0x20000000 0x2000000>;
 	flash0: flash@0 {
 		compatible = "issi,is25wp256d", "jedec,spi-nor";
+		status = "disabled";
 		size = <33554432>;
 		label = "FLASH0";
 		jedec-id = [96 60 18];

--- a/boards/riscv/hifive_unmatched/hifive_unmatched.dts
+++ b/boards/riscv/hifive_unmatched/hifive_unmatched.dts
@@ -27,12 +27,12 @@
 	current-speed = <115200>;
 };
 
+/* disabled (used by Flash ROM by default) */
 &spi0 {
-	status = "okay";
-
 	reg = <0x10040000 0x1000 0x20000000 0x2000000>;
 	flash0: flash@0 {
 		compatible = "issi,is25wp256d", "jedec,spi-nor";
+		status = "disabled";
 		size = <33554432>;
 		label = "FLASH0";
 		jedec-id = [96 60 18];

--- a/drivers/flash/flash_shell.c
+++ b/drivers/flash/flash_shell.c
@@ -14,7 +14,6 @@
 #include <stdlib.h>
 #include <string.h>
 #include <zephyr/drivers/flash.h>
-#include <soc.h>
 
 /* Buffer is only needed for bytes that follow command and offset */
 #define BUF_ARRAY_CNT (CONFIG_SHELL_ARGC_MAX - 2)

--- a/drivers/spi/Kconfig.sifive
+++ b/drivers/spi/Kconfig.sifive
@@ -3,20 +3,8 @@
 # Copyright (c) 2018 SiFive Inc.
 # SPDX-License-Identifier: Apache-2.0
 
-menuconfig SPI_SIFIVE
+config SPI_SIFIVE
 	bool "SiFive SPI controller driver"
 	depends on SOC_SERIES_RISCV_SIFIVE_FREEDOM
 	help
 	  Enable the SPI peripherals on SiFive Freedom processors
-
-config SIFIVE_SPI_0_ROM
-	bool "SPI 0 is used to access SPI Flash ROM"
-	default y
-	depends on SPI_SIFIVE
-	help
-	  If enabled, SPI 0 is reserved for accessing the SPI flash ROM and a
-	  driver interface won't be instantiated for SPI 0.
-
-	  Beware disabling this option on HiFive 1! The SPI flash ROM is where the
-	  program is stored, and if this driver initializes the interface for
-	  peripheral control the FE310 will crash on boot.

--- a/drivers/spi/spi_sifive.c
+++ b/drivers/spi/spi_sifive.c
@@ -296,24 +296,6 @@ static struct spi_driver_api spi_sifive_api = {
 			&spi_sifive_cfg_##n, \
 			POST_KERNEL, \
 			CONFIG_SPI_INIT_PRIORITY, \
-			&spi_sifive_api)
+			&spi_sifive_api);
 
-#ifndef CONFIG_SIFIVE_SPI_0_ROM
-#if DT_INST_NODE_HAS_PROP(0, label)
-
-SPI_INIT(0);
-
-#endif /* DT_INST_NODE_HAS_PROP(0, label) */
-#endif /* !CONFIG_SIFIVE_SPI_0_ROM */
-
-#if DT_INST_NODE_HAS_PROP(1, label)
-
-SPI_INIT(1);
-
-#endif /* DT_INST_NODE_HAS_PROP(1, label) */
-
-#if DT_INST_NODE_HAS_PROP(2, label)
-
-SPI_INIT(2);
-
-#endif /* DT_INST_NODE_HAS_PROP(2, label) */
+DT_INST_FOREACH_STATUS_OKAY(SPI_INIT)

--- a/dts/bindings/spi/sifive,spi0.yaml
+++ b/dts/bindings/spi/sifive,spi0.yaml
@@ -1,7 +1,12 @@
 # Copyright (c) 2018, SiFive Inc.
 # SPDX-License-Identifier: Apache-2.0
 
-description: Sifive SPI controller
+description: |
+  Sifive SPI controller.
+
+  Note: First instance of the Sifive SPI controller (spi0) must be kept
+  disabled if used to access SPI Flash ROM. Failing to do so could result in
+  crashes during boot time.
 
 compatible: "sifive,spi0"
 

--- a/include/zephyr/arch/riscv/common/linker.ld
+++ b/include/zephyr/arch/riscv/common/linker.ld
@@ -40,7 +40,7 @@
 #define ROM_BASE DT_REG_ADDR(DT_CHOSEN(zephyr_flash))
 #endif /* CONFIG_FLASH_LOAD_OFFSET */
 #define ROM_SIZE DT_REG_SIZE(DT_CHOSEN(zephyr_flash))
-#elif DT_NODE_HAS_COMPAT_STATUS(DT_CHOSEN(zephyr_flash), jedec_spi_nor, okay)
+#elif DT_NODE_HAS_COMPAT(DT_CHOSEN(zephyr_flash), jedec_spi_nor)
 /* For jedec,spi-nor we expect the spi controller to memory map the flash
  * and for that mapping to be the second register property of the spi
  * controller.

--- a/samples/drivers/flash_shell/src/main.c
+++ b/samples/drivers/flash_shell/src/main.c
@@ -13,7 +13,6 @@
 #include <zephyr/shell/shell.h>
 #include <zephyr/drivers/flash.h>
 #include <zephyr/device.h>
-#include <soc.h>
 #include <stdlib.h>
 
 LOG_MODULE_REGISTER(app);

--- a/samples/drivers/flash_shell/src/main.c
+++ b/samples/drivers/flash_shell/src/main.c
@@ -13,7 +13,6 @@
 #include <zephyr/shell/shell.h>
 #include <zephyr/drivers/flash.h>
 #include <zephyr/device.h>
-#include <stdlib.h>
 
 LOG_MODULE_REGISTER(app);
 

